### ISSUE TITLE
FIPS - Allow for selecting Openssl private key module hashing algorithm for fingerprint

### ISSUE THREAD

**Issue by jeffcpullen** [ansible-collections/community.crypto#144]
##### SUMMARY

The Openssl private key module gathers fingerprints of the the key as return data using all available hashing methods. If one of the methods is not permitted on the system due to enabling FIPS, the module fails. There is no apparent way to modify this behavior.

Previous iterations of the kubernetes metering operator (as an example) have made this work by limiting the number of available hashing algorithms [link](https://github.com/kube-reporting/metering-operator/blob/release-4.6/images/metering-ansible-operator/roles/meteringconfig/module_utils/crypto.py#L158**) but this requires blacklisting algorithms as they fall out of favor with FIPS. I'd like to propose instead allowing users to specify the fingerprint algorithm and have it run only a single default (strong) hash for the fingerprint return. If changing the default behavior to whitelist is a breaking change, just adding an option to select the hash would simplify the FIPS usability.

##### ISSUE TYPE
- Feature request

##### COMPONENT NAME

community.crypto.openssl_privatekey

##### ANSIBLE VERSION

```
ansible --version
ansible 2.9.15
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible/openshift']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.8 (default, Dec  5 2019, 15:45:45) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
```

##### CONFIGURATION

```
ansible-config dump --only-changed
DEFAULT_CALLBACK_WHITELIST(/etc/ansible/ansible.cfg) = ['timer', 'profile_roles']
DEFAULT_MODULE_PATH(/etc/ansible/ansible.cfg) = ['/usr/share/ansible/openshift']
DEFAULT_ROLES_PATH(/etc/ansible/ansible.cfg) = ['/opt/ansible/roles']

```

##### OS / ENVIRONMENT
OpenShift Operator Deployment
https://github.com/kube-reporting/metering-operator

##### STEPS TO REPRODUCE

Deploy the 4.6 Metering Operator on OpenShift 4.6 cluster

##### EXPECTED RESULTS
Metering operator creates openssl certificates

##### ACTUAL RESULTS

```
 TASK [Generate a RSA private key for the CA] ******************************** 
[0;31mAn exception occurred during task execution. To see the full traceback, use -vvv. The error was: ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS[0m
[0;31mfatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/.ansible-/tmp/ansible-tmp-1605639519.1729925-24168-234385709602156/AnsiballZ_openssl_privatekey.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/tmp/.ansible-/tmp/ansible-tmp-1605639519.1729925-24168-234385709602156/AnsiballZ_openssl_privatekey.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/tmp/.ansible-/tmp/ansible-tmp-1605639519.1729925-24168-234385709602156/AnsiballZ_openssl_privatekey.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible.modules.crypto.openssl_privatekey', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/modules/crypto/openssl_privatekey.py\", line 692, in <module>\n  File \"/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/modules/crypto/openssl_privatekey.py\", line 676, in main\n  File \"/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/modules/crypto/openssl_privatekey.py\", line 303, in generate\n  File \"/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/modules/crypto/openssl_privatekey.py\", line 545, in _get_fingerprint\n  File \"/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/module_utils/crypto.py\", line 164, in get_fingerprint_of_bytes\nValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}[0m

```

**Comment by felixfontein** [comment]
It looks like you are **not** using the modules from this collection, but the ones included in Ansible 2.9. Also, interestingly, the stacktrace you posted claims the crash happens in a comment: https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/module_utils/crypto.py#L164 To properly fix it (from the code it should also not really happen, except in case the `ValueError` is only raised a lot later), I need to know where exactly in the code this error is raised. Can you check out what's in line 164 of your Ansible 2.9 installation in module_utils/crypto.py? Thanks.

The un-mangled traceback is:
```
Traceback (most recent call last):
  File "/tmp/.ansible-/tmp/ansible-tmp-1605639519.1729925-24168-234385709602156/AnsiballZ_openssl_privatekey.py", line 102, in <module>
    _ansiballz_main()
  File "/tmp/.ansible-/tmp/ansible-tmp-1605639519.1729925-24168-234385709602156/AnsiballZ_openssl_privatekey.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/tmp/.ansible-/tmp/ansible-tmp-1605639519.1729925-24168-234385709602156/AnsiballZ_openssl_privatekey.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible.modules.crypto.openssl_privatekey', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/modules/crypto/openssl_privatekey.py", line 692, in <module>
  File "/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/modules/crypto/openssl_privatekey.py", line 676, in main
  File "/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/modules/crypto/openssl_privatekey.py", line 303, in generate
  File "/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/modules/crypto/openssl_privatekey.py", line 545, in _get_fingerprint
  File "/tmp/ansible_openssl_privatekey_payload_rg2zb9st/ansible_openssl_privatekey_payload.zip/ansible/module_utils/crypto.py", line 164, in get_fingerprint_of_bytes
ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS
```

**Comment by jeffcpullen** [comment]
@felixfontein Ok that makes a lot more sense. The crypto module is actually being packaged with the [metering ansible operator](https://github.com/kube-reporting/metering-operator/blob/release-4.6/images/metering-ansible-operator/roles/meteringconfig/module_utils/crypto.py#L158**)

It looks like you committed some code in March specifically to address this issue to the Ansible 2.9, but it wasn't included in the ansible-metering-operator. Thank you for pointing me in the right direction. I'll move my issue over to that project.

**Comment by felixfontein** [comment]
@jeffcpullen ah! If you're looking for the references for that fix (though I guess you found at least some of them already), see ansible/ansible#67515 for merging the fix into Ansible's devel branch, and ansible/ansible#67519 for the backport to the stable-2.9 branch (which put that fix into Ansible 2.9). The fix should have first shown up in Ansible 2.9.6, released on March 5, 2020.

**Comment by jeffcpullen** [comment]
Moved this over to the right place. Thank you @felixfontein for pointing me in the right direction!

**Comment by MarkusTeufelberger** [comment]
Seems to come from https://github.com/kube-reporting/metering-operator/pull/1088 - sadly they don't seem to have upstreamed their fix or even opened an issue in Ansible about this. :-(

One month after their vendored fix https://github.com/ansible/ansible/issues/67213 was opened, fixed 2 weeks later and released a month in total after the report.

**Comment by MarkusTeufelberger** [comment]
I checked if they did anything else there, but it seems they just vendored the module_utils for this single fix, so they could just require Ansible >=2.9.6 and remove their version.

**Comment by felixfontein** [comment]
@MarkusTeufelberger thanks for tracking down the reports. If that initial report on January 4th would have been reported to upstream immediately (and labelled correctly), the fix would have probably been in Ansible 2.9.4 (released January 15th).
